// diff/lib/patch/line-endings.js
/*istanbul ignore start*/
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isUnix = isUnix;
exports.isWin = isWin;
exports.unixToWin = unixToWin;
exports.winToUnix = winToUnix;
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
/*istanbul ignore end*/
function unixToWin(patch) {
  if (Array.isArray(patch)) {
    return patch.map(unixToWin);
  }
  return (
    /*istanbul ignore start*/
    _objectSpread(_objectSpread({},
    /*istanbul ignore end*/
    patch), {}, {
      hunks: patch.hunks.map(function (hunk)
      /*istanbul ignore start*/
      {
        return _objectSpread(_objectSpread({},
        /*istanbul ignore end*/
        hunk), {}, {
          lines: hunk.lines.map(function (line, i)
          /*istanbul ignore start*/
          {
            var _hunk$lines;
            return (
              /*istanbul ignore end*/
              line.startsWith('\\') || line.endsWith('\r') ||
              /*istanbul ignore start*/
              (_hunk$lines =
              /*istanbul ignore end*/
              hunk.lines[i + 1]) !== null && _hunk$lines !== void 0 &&
              /*istanbul ignore start*/
              _hunk$lines
              /*istanbul ignore end*/
              .startsWith('\\') ? line : line + '\r'
            );
          })
        });
      })
    })
  );
}
function winToUnix(patch) {
  if (Array.isArray(patch)) {
    return patch.map(winToUnix);
  }
  return (
    /*istanbul ignore start*/
    _objectSpread(_objectSpread({},
    /*istanbul ignore end*/
    patch), {}, {
      hunks: patch.hunks.map(function (hunk)
      /*istanbul ignore start*/
      {
        return _objectSpread(_objectSpread({},
        /*istanbul ignore end*/
        hunk), {}, {
          lines: hunk.lines.map(function (line)
          /*istanbul ignore start*/
          {
            return (
              /*istanbul ignore end*/
              line.endsWith('\r') ? line.substring(0, line.length - 1) : line
            );
          })
        });
      })
    })
  );
}

/**
 * Returns true if the patch consistently uses Unix line endings (or only involves one line and has
 * no line endings).
 */
function isUnix(patch) {
  if (!Array.isArray(patch)) {
    patch = [patch];
  }
  return !patch.some(function (index)
  /*istanbul ignore start*/
  {
    return (
      /*istanbul ignore end*/
      index.hunks.some(function (hunk)
      /*istanbul ignore start*/
      {
        return (
          /*istanbul ignore end*/
          hunk.lines.some(function (line)
          /*istanbul ignore start*/
          {
            return (
              /*istanbul ignore end*/
              !line.startsWith('\\') && line.endsWith('\r')
            );
          })
        );
      })
    );
  });
}

/**
 * Returns true if the patch uses Windows line endings and only Windows line endings.
 */
function isWin(patch) {
  if (!Array.isArray(patch)) {
    patch = [patch];
  }
  return patch.some(function (index)
  /*istanbul ignore start*/
  {
    return (
      /*istanbul ignore end*/
      index.hunks.some(function (hunk)
      /*istanbul ignore start*/
      {
        return (
          /*istanbul ignore end*/
          hunk.lines.some(function (line)
          /*istanbul ignore start*/
          {
            return (
              /*istanbul ignore end*/
              line.endsWith('\r')
            );
          })
        );
      })
    );
  }) && patch.every(function (index)
  /*istanbul ignore start*/
  {
    return (
      /*istanbul ignore end*/
      index.hunks.every(function (hunk)
      /*istanbul ignore start*/
      {
        return (
          /*istanbul ignore end*/
          hunk.lines.every(function (line, i)
          /*istanbul ignore start*/
          {
            var _hunk$lines2;
            return (
              /*istanbul ignore end*/
              line.startsWith('\\') || line.endsWith('\r') ||
              /*istanbul ignore start*/
              ((_hunk$lines2 =
              /*istanbul ignore end*/
              hunk.lines[i + 1]) === null || _hunk$lines2 === void 0 ? void 0 :
              /*istanbul ignore start*/
              _hunk$lines2
              /*istanbul ignore end*/
              .startsWith('\\'))
            );
          })
        );
      })
    );
  });
}
