diff -u fo.ORIG/param.xsl fo/param.xsl
--- fo.ORIG/param.xsl	2005-11-30 19:22:15.992409173 +1300
+++ fo/param.xsl	2005-11-30 11:25:40.000000000 +1300
@@ -669,6 +669,7 @@
 </xsl:attribute-set>
 <xsl:param name="segmentedlist.as.table" select="0"/>
 <xsl:param name="shade.verbatim" select="0"/>
+<xsl:param name="shade.literallayout" select="0"/>
 
 <xsl:attribute-set name="shade.verbatim.style">
   <xsl:attribute name="background-color">#E0E0E0</xsl:attribute>
diff -u fo.ORIG/verbatim.xsl fo/verbatim.xsl
--- fo.ORIG/verbatim.xsl	2005-11-30 19:22:15.944415115 +1300
+++ fo/verbatim.xsl	2005-11-30 11:26:12.000000000 +1300
@@ -105,7 +105,7 @@
   <xsl:choose>
     <xsl:when test="@class='monospaced'">
       <xsl:choose>
-        <xsl:when test="$shade.verbatim != 0">
+        <xsl:when test="$shade.literallayout != 0">
           <fo:block id="{$id}"
                     white-space-collapse='false'
                     white-space-treatment='preserve'
@@ -128,7 +128,7 @@
     </xsl:when>
     <xsl:otherwise>
       <xsl:choose>
-        <xsl:when test="$shade.verbatim != 0">
+        <xsl:when test="$shade.literallayout != 0">
           <fo:block id="{$id}"
                     wrap-option='no-wrap'
                     white-space-collapse='false'
