--- openssl-1.0.0f-origin/crypto/cryptlib.c	2011-06-22 23:39:00.000000000 +0800
+++ openssl-1.0.0f/crypto/cryptlib.c	2012-01-19 02:17:50.261681856 +0800
@@ -690,7 +690,7 @@
 unsigned long *OPENSSL_ia32cap_loc(void) { return NULL; }
 #endif
 int OPENSSL_NONPIC_relocated = 0;
-#if !defined(OPENSSL_CPUID_SETUP) && !defined(OPENSSL_CPUID_OBJ)
+#if !defined(OPENSSL_CPUID_SETUP)
 void OPENSSL_cpuid_setup(void) {}
 #endif
 
