commit 90641d2d6d251d014815e1703157f0cd0a6d5ea1
Author: Adam Langley <agl@chromium.org>
Date:   Thu Nov 4 16:02:17 2010 -0400

    missing_stddef.patch

diff --git a/crypto/md2/md2.h b/crypto/md2/md2.h
index d59c9f2..75446fd 100644
--- a/crypto/md2/md2.h
+++ b/crypto/md2/md2.h
@@ -72,6 +72,8 @@
 extern "C" {
 #endif
 
+#include <stddef.h>
+
 typedef struct MD2state_st
 	{
 	unsigned int num;
diff --git a/crypto/md4/md4.h b/crypto/md4/md4.h
index ba1fe4a..d9b2717 100644
--- a/crypto/md4/md4.h
+++ b/crypto/md4/md4.h
@@ -66,6 +66,8 @@
 extern "C" {
 #endif
 
+#include <stddef.h>
+
 #ifdef OPENSSL_NO_MD4
 #error MD4 is disabled.
 #endif
diff --git a/crypto/md5/md5.h b/crypto/md5/md5.h
index 0761f84..1a51bcf 100644
--- a/crypto/md5/md5.h
+++ b/crypto/md5/md5.h
@@ -66,6 +66,8 @@
 extern "C" {
 #endif
 
+#include <stddef.h>
+
 #ifdef OPENSSL_NO_MD5
 #error MD5 is disabled.
 #endif
diff --git a/crypto/ripemd/ripemd.h b/crypto/ripemd/ripemd.h
index 3b6d043..70a9637 100644
--- a/crypto/ripemd/ripemd.h
+++ b/crypto/ripemd/ripemd.h
@@ -66,6 +66,8 @@
 extern "C" {
 #endif
 
+#include <stddef.h>
+
 #ifdef OPENSSL_NO_RIPEMD
 #error RIPEMD is disabled.
 #endif
diff --git a/crypto/sha/sha.h b/crypto/sha/sha.h
index 47a2c29..1f6ad37 100644
--- a/crypto/sha/sha.h
+++ b/crypto/sha/sha.h
@@ -66,6 +66,8 @@
 extern "C" {
 #endif
 
+#include <stddef.h>
+
 #if defined(OPENSSL_NO_SHA) || (defined(OPENSSL_NO_SHA0) && defined(OPENSSL_NO_SHA1))
 #error SHA is disabled.
 #endif
